### PR TITLE
Adding search page

### PR DESCRIPTION
**Pull request by OfirBachar**
This PR was created by @idoyek and me (@OfirBachar ).
This is the skeleton of adding a search professionals page.
Logic and UI improvements will be added in the future soon.

Updating Pro4U/urls.py file
Adding SearchHistory/urls.py file
Adding SearchHistory/views.py file
Adding SearchHistory/templates/html/search.html file

Resolves issue #94

**Take a look!**

image

### CODE REVIEW

File: SearchHistory/views.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from django.shortcuts import render
+# from django.shortcuts import redirect
+# from .models import Message
+# from .forms import MessageForm
```
Comment (patrisiakaplun) [inline SearchHistory/views.py:4]: If these lines are not needed, please delete them

File: SearchHistory/templates/html/search.html [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            font-size: 16px;
+            color: #666;
+        }
+
+        .professional-grid {
+            display: grid;
+            grid-template-columns: repeat(2, 1fr);
+            grid-gap: 20px;
+        }
+    </style>
```
Comment (patrisiakaplun) [inline SearchHistory/templates/html/search.html:107]: It would be better to put all the style in a separate css file

**Comment by patrisiakaplun** [comment]
If you are going to attach the tests and the logic in a separate PR, please open an issue for it

File: SearchHistory/templates/html/search.html (hunk 2 above)
Comment (OfirBachar) [reply SearchHistory/templates/html/search.html]: Done

File: SearchHistory/views.py (hunk 1 above)
Comment (OfirBachar) [reply SearchHistory/views.py]: Done

**Comment by OfirBachar** [comment]
> If you are going to attach the tests and the logic in a separate PR, please open an issue for it

It will be in the same PR.

**Review by idosinger98 — COMMENTED** [review]
Can you explain what the purpose of PR is? I don't see anything in the view.py file and it's unclear how it's supposed to work

**Comment by Yarboa** [comment]
> Can you explain what the purpose of PR is? I don't see anything in the view.py file and it's unclear how it's supposed to work

@OfirBachar ^^ please reply, 
I would also add, that this PR should have tests, 
Otherwise it will not move forward,
Please suggest a minimum set of DataFlow to UI

**Review by idosinger98 — COMMENTED** [review]
I don't see the test.py file

**Comment by OfirBachar** [comment]
> I don't see the test.py file

It's in progress, I wrote in the description that tests will be added in the future soon, working on it.

**Comment by idoyek** [comment]
I wrote some tests for this PR, feel free to add them @OfirBachar :
[test_search_page.zip](https://github.com/redhat-beyond/Pro4U/files/11485292/test_search_page.zip)

**Comment by OfirBachar** [comment]
> > Can you explain what the purpose of PR is? I don't see anything in the view.py file and it's unclear how it's supposed to work
> 
> @OfirBachar ^^ please reply, I would also add, that this PR should have tests, Otherwise it will not move forward, Please suggest a minimum set of DataFlow to UI

Of course! the views file was updated. in the beginning, it was just a skeleton, now it is much more. 
@idosinger98 replay for your question. take a look at the updated views file.

File: SearchHistory/views.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    elif request.method == 'POST':
+        professionals = Professional.objects.all()
+
+        professional_id = request.POST.get("professional_id", "")
+        profession = request.POST.get("profession", "")
+        first_name = request.POST.get("first_name", "")
+        last_name = request.POST.get("last_name", "")
+        city = request.POST.get("city", "")
+        # professional_id_searched = request.POST.get("professional_id_searched", "")
```
Comment (Yarboa) [inline SearchHistory/views.py:31]: Please remove

File: SearchHistory/views.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        context = {
+            'professionals': professionals,
+            'last_searches': last_searches,
+        }
+
+        return render(request, 'html/search.html', context=context)
+
+
+@permission_required('your_app.create_search_history')
```
Comment (Yarboa) [inline SearchHistory/views.py:53]: Why is it needed? 
https://docs.djangoproject.com/en/4.2/topics/auth/default/#the-permission-required-decorator

File: SearchHistory/models.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 from django.utils import timezone
 from account.models.professional import Professional
 from account.models.client import Client
-from django.db.models import Max
+# from django.db.models import Max
```
Comment (Yarboa) [inline SearchHistory/models.py:5]: Is it needed?

File: SearchHistory/templates/html/search.html [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    </div>
+</div>
+{% endblock %}
+
+{% block scripts %}
+<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.bundle.min.js"
+    integrity="sha384-s9lT2w4UDTe8rDwWyTtHofbFH3fa6zvVnfwBlyUtdS+mxL/M9aQjvI99vHYMtoVw"
+    crossorigin="anonymous"></script>
+
+<!-- <script>
```
Comment (Yarboa) [inline SearchHistory/templates/html/search.html:73]: Is it needed?

File: SearchHistory/tests.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 from SearchHistory.models import SearchHistory
 import pytest
-from django.db.models import Max
+# from django.db.models import Max
```
Comment (Yarboa) [inline SearchHistory/tests.py:3]: Is it needed?

File: SearchHistory/views.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+def create_search_history(request):
+    if request.method == 'POST':
+        data = json.loads(request.body)
+        client_id = data['client_id']
+        professional_id = data['professional_id']
+
+        SearchHistory.create_new_search_history(client_id=client_id, professional_id=professional_id)
+        # Perform the necessary logic with the professional ID
```
Comment (Yarboa) [inline SearchHistory/views.py:61]: Is it needed?

File: SearchHistory/test_search_page.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    url = reverse('search history', args=[CLIENT_ID])
+
+    response = client.post(url, {'professional_id': professional.professional_id})
+
+    assert response.status_code == 200
+    assert 'html/search.html' in response.templates[0].name
+
+    professionals = response.context['professionals']
+    assert len(professionals) == 1
+    assert professionals[0].professional_id == professional.professional_id
```
Comment (idoyek) [inline SearchHistory/test_search_page.py:37]: Please add also these lines:
    assert professionals[0].profession == professional.profession
    assert professionals[0].profile_id.user_id.first_name == professional.profile_id.user_id.first_name
    assert professionals[0].profile_id.user_id.last_name == professional.profile_id.user_id.last_name
    assert professionals[0].profile_id.city == professional.profile_id.city

File: SearchHistory/views.py (hunk 3 above)
Comment (OfirBachar) [reply SearchHistory/views.py]: Fixed

File: SearchHistory/test_search_page.py (hunk 9 above)
Comment (OfirBachar) [reply SearchHistory/test_search_page.py:40]: Done

File: SearchHistory/models.py (hunk 5 above)
Comment (idoyek) [reply SearchHistory/models.py]: Resolved

File: SearchHistory/templates/html/search.html (hunk 6 above)
Comment (idoyek) [reply SearchHistory/templates/html/search.html]: Resolved

File: SearchHistory/tests.py (hunk 7 above)
Comment (idoyek) [reply SearchHistory/tests.py]: Resolved

File: SearchHistory/views.py (hunk 8 above)
Comment (idoyek) [reply SearchHistory/views.py]: Resolved

**Review by idoyek — APPROVED** [review]
LGTM

**Review by idosinger98 — COMMENTED** [review]
What happens if you click on one of the professionals who are in the last 5 search history?

**Comment by idoyek** [comment]
> What happens if you click on one of the professionals who are in the last 5 search history?

@idosinger98 we don’t want to show duplicate professionals, so the one that was already showed  will go to be the first one now in the top, because he is the last one that was searched

**Comment by OfirBachar** [comment]
> What happens if you click on one of the professionals who are in the last 5 search history?

it will update the date of the search history with that professional to be the new last search. distinct style.
take a look at the crearte_new_search_history method.

**Review by idosinger98 — APPROVED** [review]
LGTM

**Comment by patrisiakaplun** [comment]
https://github.com/redhat-beyond/Pro4U/pull/90#discussion_r1193882260 ? (:

**Review by Yarboa — APPROVED** [review]
LGTM

**Review by guybeckenstein — APPROVED** [review]
LGTM

**Review by idosinger98 — CHANGES_REQUESTED** [review]
Why did you reupload the approved files?

Please remove it

**Comment by idoyek** [comment]
> Why did you reupload the approved files?
> 
> Please remove it

Done

**Review by idosinger98 — APPROVED** [review]
LGTM
